Libs/DICOM/Core/ctkDICOMDatabase.cpp

```diff
@@ -1444,6 +1447,19 @@ QString ctkDICOMDatabasePrivate::internalPathFromAbsolute(const QString& filenam
   return ctk::internalPathFromAbsolute(filename, q->databaseDirectory());
 }
 
+//------------------------------------------------------------------------------
+QString ctkDICOMDatabasePrivate::absolutePathFromInternalIfFile(const QString& filename)
+{
+  Q_Q(ctkDICOMDatabase);
+  QString returnFileName = filename;
+  if (QUrl(filename).scheme().isEmpty()) // local path, not url
```

Comment:
What do we allow to be stored in the "Filename" database field?

We could say that it is always a URL, but unfortunately it is not so simple, because based on the standard definition of a URL, because an absolute windows path appears as a URL (e.g., `x:/some/path`). For example:

```
>>> url = qt.QUrl('x:/some/path')
>>> url.scheme()
'x'
>>> url.path()
'/some/path'
```

I see two options:

### Option A: Always store a URL

We can store relative local file paths without any change the same way as before. For example: `DICOM/some/file.dcm`

We can store remote URLs without any change. For example: `https://example.com/dicom/some/file.dcm`

We need to change how we store Windows absolute paths. We cannot store them as `x:/some/file.dcm` but we must store it as `file:///x:/some/file.dcm`.

```
>>> url = qt.QUrl.fromLocalFile('x:/some/file.dcm')
>>> url.toString()
'file:///x:/some/file.dcm'
>>> url.scheme()
'file'
>>> url.path()
'/x:/some/file.dcm'
```

### Option B: Store as local file path or URL

Detect if the string is a local file path or a URL using a custom rule: if `schema` is `http` or `https` then it is a URL; otherwise it is a local file path.

---

Option B is not standard, therefore we would need o document the API carefully. We would also need to find a good name to this string that can be "URL or local path" (it would not be precise to call it a URL, because `x:/some/thing` must not be interpreted as a URL). Still, probably it would be simpler than forcing the `file:///` prefix for all Windows absolute paths.

Reply:
> What do we allow to be stored in the "Filename" database field?

It's just a string. (`VARCHAR(1024) NOT NULL` in sql).



> an absolute windows path appears as a URL

Good point - we'll need to find a way to handle this.  It may not be an issue in practice right now, since it's already an absolute path the logic actually works, but something more robust and clear would be better.

A better solution could be to improve the database to clarify what the field is intended to store.  Perhaps an enumerated type for `LocalFilePath`, `URL`, `CustomLocator` could be added.

If we are okay with updating the schema, which in general is accommodated by the design we could.  But I prefer to do that as rarely as feasible.

Reply:
Maybe having a new `URL` column in addition to the current `LocalFilePath` column would be the most useful. If a file is downloaded and available locally we would fill the `LocalFilePath`. If the local cache uses too much disk space then we could remove local files and we would still know how to retrieve them again.

I like the custom schema idea (`cget://ipaddress/instanceuid`). For DICOM (both DIMSE and DICOMweb) it would be probably better to rely on the DICOM host list to store all the communicaton parameters (dicomweb/dimse-cget/dimse-cmove protocol, calling/called aetitle, port number, access token, proxy server, etc.) instead of trying to squeeze all these into a URL. So, the URL could simply look like this `dicom://bwhpacs01/instanceuid` (bwhpacs01 entry in the host list would tell its IP and that it communicates using DICOMweb). We could still have simple `http` schema and any custom schema (e.g., if you want to use a proprietary API to retrieve data from a cloud provider).

Updating the database schema should be tolerable, as the added value would justify the inconvenience.

Reply:
Those are reasonable ideas.  There's also the case, at least in the AWS back end and some of the dicom plugins like RT, where we'll need to construct a binary file from metadata, and possibly frame data, for compatibility with existing assumptions.  We could cache those binaries and keep track of both the local file path and original locator.

I wonder if there are other things to change in the schema while we are at it.  One nitpicky inconsistency is that we don't use standard dicom terms in the fields.  Like DICOM `PatientName` is `PatiensName` in the schema, but I'm not sure if that's ever worth fixing.